src/target/llvm.rs

```diff
@@ -0,0 +1,91 @@
+use std::borrow::Cow;
+
+use super::TargetInfo;
+
+impl TargetInfo {
+    /// The versioned LLVM/Clang target triple.
+    pub(crate) fn versioned_llvm_target(&self, version: Option<&str>) -> Cow<'_, str> {
+        if let Some(version) = version {
+            // Only support versioned Apple targets for now.
+            assert_eq!(self.vendor, "apple");
+
+            let mut components = self.unversioned_llvm_target.split("-");
+            let arch = components.next().expect("llvm_target should have arch");
+            let vendor = components.next().expect("llvm_target should have vendor");
+            let os = components.next().expect("LLVM target should have os");
+            let environment = components.next();
+            assert_eq!(components.next(), None, "too many LLVM target components");
+
+            Cow::Owned(if let Some(env) = environment {
+                format!("{arch}-{vendor}-{os}{version}-{env}")
+            } else {
+                format!("{arch}-{vendor}-{os}{version}")
+            })
+        } else {
+            Cow::Borrowed(&self.unversioned_llvm_target)
+        }
+    }
+}
+
+/// Rust and Clang don't really agree on naming, so do a best-effort
+/// conversion to support out-of-tree / custom target-spec targets.
+pub(crate) fn guess_llvm_target_triple(
+    full_arch: &str,
+    vendor: &str,
+    os: &str,
+    env: &str,
+    abi: &str,
+) -> String {
+    let arch = match full_arch {
+        riscv32 if riscv32.starts_with("riscv32") => "riscv32",
+        riscv64 if riscv64.starts_with("riscv64") => "riscv64",
+        arch => arch,
+    };
+    let os = match os {
+        "darwin" => "macosx",
+        "visionos" => "xros",
+        "uefi" => "windows",
+        os => os,
+    };
+    let env = match env {
+        "newlib" | "nto70" | "nto71" | "ohos" | "p1" | "p2" | "relibc" | "sgx" | "uclibc" => "",
+        abi => abi,
```

Comment:
Nit naming convention:

```suggestion
        env => env,
```